### PR TITLE
Enable coverage for eval.

### PR DESCRIPTION
**Pull request by ioquatix**
```
make test-all TESTS=../test/coverage/test_coverage.rb
```

I've added a test and it appears to be working, but some other test segfaults and I don't yet understand why.

### CODE REVIEW

**Comment by ioquatix** [comment]
```
  |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(ruby_sized_xfree.part.0+0x1e) [0x5595c07ba39e]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(obj_free+0xe7) [0x5595c07bc667]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(gc_sweep_step+0x1e7) [0x5595c07c4527]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(gc_sweep+0x17a) [0x5595c07c517a]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(gc_start+0xf0f) [0x5595c07cbbaf]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(newobj_alloc+0x5ba) [0x5595c07cf76a]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(rb_wb_protected_newobj_of+0x63) [0x5595c07d0433]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(str_new0.constprop.0+0x40) [0x5595c08fe570]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(rb_str_subseq+0xa7) [0x5595c090e6f7]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(nextline+0x57) [0x5595c0854837]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(parser_yylex+0x125) [0x5595c08651f5]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(ruby_yyparse+0x780) [0x5595c086e7a0]",
   |  "/home/samuel/Projects/ioquatix/ruby/build/ruby(yycompile0+0x100) [0x5595c0880240]",
```

File: iseq.c [hunk 1]
```diff
@@ -928,6 +927,18 @@ rb_iseq_new_main(const rb_ast_body_t *ast, VALUE path, VALUE realpath, const rb_
 rb_iseq_t *
 rb_iseq_new_eval(const rb_ast_body_t *ast, VALUE name, VALUE path, VALUE realpath, VALUE first_lineno, const rb_iseq_t *parent, int isolated_depth)
 {
+    VALUE coverages = rb_get_coverages();
+    if (RTEST(coverages) && RTEST(path) && !RTEST(rb_hash_has_key(coverages, path))) {
+        int line_offset = RB_NUM2INT(first_lineno) - 1;
+        int line_count = line_offset + ast_line_count(ast);
+
+        if (line_count >= 0) {
```
Comment (mame) [inline iseq.c:935]: @ioquatix Sorry to review after the merge, but this part is almost a code clone of the beginning of rb_iseq_new_top. Can you refactor it out to a function?

Comment (ioquatix) [reply iseq.c:935]: Sure, no problem.

File: spec/ruby/library/coverage/result_spec.rb [hunk 2]
```diff
@@ -91,15 +91,31 @@
...
+      result.should == {
+          @eval_code_file => [
+              1, nil, 1, nil, 1, nil, nil, nil, nil, nil, 1
+          ]
+      }
+    end
+  end
+
+  ruby_version_is '3.2' do
+    it 'returns the correct results when eval is used' do
```
Comment (eregon) [inline spec/ruby/library/coverage/result_spec.rb:109]: It's quite difficult to notice the difference here.
Could you make the specs clearer?
For example the spec description could say in 3.1 it does not include eval lines and in 3.2 it does

Comment (ioquatix) [reply spec/ruby/library/coverage/result_spec.rb:109]: Is this just a matter of changing the test names?

Comment (eregon) [reply spec/ruby/library/coverage/result_spec.rb:109]: Yes, mostly the test names.

**Pull request merged**